extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,639 @@
+    const iframeBox = {
+      top: ampAdBox.top + heightOffset,
+      bottom: ampAdBox.bottom - heightOffset,
+      left: ampAdBox.left + widthOffset,
+      right: ampAdBox.right - widthOffset,
+      height: this.creativeSize_.height,
+      width: this.creativeSize_.width,
+    };
+    return this.formatGeom_(iframeBox);
+  }
+
+  /**
+   * Registers this instance as the host API for the current sentinel.
+   * If the global safeframe listener has not yet been created, it creates
+   * that as well.
+   */
+  registerSafeframeHost() {
+    dev().assert(this.sentinel_);
+    safeframeHosts[this.sentinel_] = safeframeHosts[this.sentinel_] || this;
+    if (!safeframeListenerCreated_) {
+      safeframeListenerCreated_ = true;
+      this.win_.addEventListener('message', safeframeListener, false);
+    }
+  }
+
+  /**
+   * Sends initial connection message to the safeframe to finish initialization.
+   * Also initializes the sending of geometry update messages to the frame.
+   * @param {string} channel
+   */
+  connectMessagingChannel(channel) {
+    // Set the iframe here, because when class is first created the iframe
+    // element does not yet exist on this.baseInstance_. The first time
+    // we receive a message we know that it now exists.
+    dev().assert(this.baseInstance_.iframe);
+    this.iframe_ = this.baseInstance_.iframe;
+    this.channel = channel;
+    this.setupGeom_();
+    this.sendMessage_({
+      'message': 'connect',
+      'c': this.channel,
+    }, '');
+  }
+
+  /**
+   * Creates IntersectionObserver instance for this SafeframeAPI instance.
+   * We utilize the existing IntersectionObserver class, by passing in this
+   * class for IO to use instead of SubscriptionApi for sending its update
+   * messages. The method 'send' below is triggered by IO every time that
+   * an update occurs.
+   * @private
+   */
+  setupGeom_() {
+    dev().assert(this.iframe_.contentWindow,
+        'Frame contentWindow unavailable.');
+    this.viewport_.onScroll(this.maybeUpdateGeometry_.bind(this));
+    this.viewport_.onChanged(this.maybeUpdateGeometry_.bind(this));
+    this.updateGeometry_();
+  }
+
+  /**
+   * Attempts to call updateGeometry_ to send a geometry update into the
+   * safeframe. If it has been less than 1 second since the last update
+   * and there is still an active delay, instead it registers that
+   * a geometry update is required.
+   * @private
+   */
+  maybeUpdateGeometry_() {
+    this.sendPositionUpdate_ = true;
+    if (!this.delay_) {
+      this.updateGeometry_();
+    }
+  }
+
+  /**
+   * Sends a geometry update message into the safeframe, and sets a timer
+   * to prevent another update being sent for 1 second. When the timer is
+   * up, if this.sendPositionUpdate_ is set, then it sends another update.
+   * @private
+   */
+  updateGeometry_() {
+    if (!this.iframe_) {
+      return;
+    }
+    this.delay_ = Services.timerFor(this.win_).promise(1000).then(() => {
```

Comment:
Would `throttle` function achieve the same effect?

Reply:
switched